ranking_evaluator.py

```diff
@@ -0,0 +1,132 @@
+__copyright__ = "Copyright (c) 2021 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+from typing import Optional, Union, List, Dict, Tuple
+
+from jina.logging.logger import JinaLogger
+from jina import DocumentArray, Executor, requests, Document
+from jina.types.arrays.traversable import TraversableSequence
+
+from metrics import precision, recall, reciprocal_rank, average_precision, fscore, ndcg
+
+
+class RankingEvaluator(Executor):
+    class DocGroundtruthPair:
+        def __init__(self, doc: 'Document', groundtruth: 'Document'):
+            self.pair = (doc, groundtruth)
+
+        def __getitem__(self, item):
+            return self.pair[item]
+
+        @property
+        def doc(self):
+            return self[0]
+
+        @property
+        def groundtruth(self):
+            return self[1]
+
+        @property
+        def matches(self):
+            pairs = []
+            for doc, groundtruth in zip(self.doc.matches, self.groundtruth.matches):
+                pairs.append(RankingEvaluator.DocGroundtruthPair(doc, groundtruth))
+            return RankingEvaluator.DocumentGroundtruthArray(pairs)
+
+        @property
+        def chunks(self):
+            assert len(self.doc.chunks) == len(self.groundtruth.chunks)
+            pairs = []
+            for doc, groundtruth in zip(self.doc.chunks, self.groundtruth.chunks):
+                pairs.append(RankingEvaluator.DocGroundtruthPair(doc, groundtruth))
+            return RankingEvaluator.DocumentGroundtruthArray(pairs)
+
+    class DocumentGroundtruthArray(TraversableSequence):
+        def __init__(self, pairs):
+            self._pairs = pairs
+
+        @property
+        def matches(self):
+            pairs = []
+            for doc, groundtruth in zip(self.doc.matches, self.groundtruth.matches):
+                pairs.append(RankingEvaluator.DocGroundtruthPair(doc, groundtruth))
+            return RankingEvaluator.DocumentGroundtruthArray(pairs)
```

Comment:
can we unify `DocGroundtruthPair ` and `DocumentGroundtruthArray `, such as `DocGroundtruthArray`

Reply:
Done! Good catch!